feat(skeleton-loader): add skeleton loader package (#1641)
twilio-labs/paste: packages/paste-website/src/pages/components/skeleton-loader/index.mdx

export const pageQuery = graphql`
  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/skeleton-loader"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(fields: {slug: {eq: "/components/skeleton-loader/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
      }
      headings {
        depth
        value
      }
    }
    allAirtable(filter: {data: {Feature: {eq: "Skeleton Loader"}}}) {
      edges {
        node {
          data {
            Documentation
            Figma
            Design_committee_review
            Engineer_committee_review
            Code
            status
          }
        }
      }
    }
  }
`;

<ComponentHeader
  name="Skeleton Loader"
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/skeleton-loader"
  storybookUrl="/?path=/story/components-skeleton-loader--default"
  data={props.data.allPasteComponent.edges}
  packageStatus={props.data.allAirtable.edges}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

> ## Guidelines

Reviewer: Similarly to Alerts, we're going to start adding a basic default example to the top of every component page so you see the component up front.

You could probably just get away with reusing the default example.

### About Skeleton Loader

The Skeleton Loader displays a placeholder representing the page's content before the page is ready to display. This component is used to improve [perceived performance](https://developer.mozilla.org/en-US/docs/Learn/Performance/Perceived_performance). This tends to improves user experience by reducing load time frustration and making the page feel more responsive.

This loader is an alternative to the [Spinner component](/components/spinner). Rather than showing an abstract spinner, it frames the content of what is to come which creates anticipation.

The component is designed to be used in place of the content being loaded. For example:

<LivePreview scope={{SkeletonLoader, Text}} noInline>
  {CompositionExample}
</LivePreview>

### Accessibility

- The animation can be disabled using the `disabled` prop, or by enabling the `prefers-reduced-motion` setting at the OS level.
- Skeleton Loader has the `aria-busy` attribute set to `true` by default. This is to indicate that the content is loading. This is reset to `false` when `disabled` is `true`.

## Examples

### Default Skeleton Loader

> The default Skeleton Loader uses the `sizeIcon20` token for the `height`, and `borderRadius20` for the `border-radius`. The `width` will be 100% unless a different `width` is specified.

Reviewer 2: It'd be helpful to say what someone might use the default Skeleton Loader for (line of text, maybe?), and that way people can use that size as a reference point for scaling up/down their sizing.

  {
    allPasteComponent(filter: {name: {eq: "@twilio-paste/skeleton-loader"}}) {
      edges {
        node {
          name
          description
          status
          version
        }
      }
    }
    mdx(fields: {slug: {eq: "/components/skeleton-loader/"}}) {
      fileAbsolutePath
      frontmatter {
        slug
        title
      }
      headings {
        depth
        value
      }
    }
    allAirtable(filter: {data: {Feature: {eq: "Skeleton Loader"}}}) {
      edges {
        node {
          data {
            Documentation
            Figma
            Design_committee_review
            Engineer_committee_review
            Code
            status
          }
        }
      }
    }
  }
`;

<ComponentHeader
  name="Skeleton Loader"
  categoryRoute={SidebarCategoryRoutes.COMPONENTS}
  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/skeleton-loader"
  storybookUrl="/?path=/story/components-skeleton-loader--default"
  data={props.data.allPasteComponent.edges}
  packageStatus={props.data.allAirtable.edges}
/>

---

<contentwrapper>

<PageAside data={props.data.mdx} />

<content>

## Guidelines

### About Skeleton Loader

> Skeleton Loader displays a placeholder of a page's content before the data gets loaded. This helps reduce load-time frustration by making it feel like loading is happening immediately.
> 
> It's an alternative to the traditionaly spinner loading methods. Rather than showing an abstract spinner, it creates anticipation of whats to come and reduces congitive load.

Reviewer 3: ```suggestion
The Skeleton Loader displays a placeholder representing the page's content before the page is ready to display. This component is used to improve [perceived performance](https://developer.mozilla.org/en-US/docs/Learn/Performance/Perceived_performance). This tends to improves user experience by reducing load time frustration and making the page feel more responsive.

This loader is an alternative to the [Spinner component](/components/spinner). Rather than showing an abstract spinner, it frames the content of what is to come which creates anticipation.
```

Reviewer 3: Took a stab at clarifying and mentioning perceived performance.

> The Skeleton Loader displays a placeholder representing the page's content before the page is ready to display. This component is used to improve [perceived performance](https://developer.mozilla.org/en-US/docs/Learn/Performance/Perceived_performance). This tends to improves user experience by reducing load time frustration and making the page feel more responsive.

Reviewer 4: Linking to more details about Perceived Performance is a really good idea!

> The Skeleton Loader displays a placeholder representing the page's content before the page is ready to display. This component is used to improve [perceived performance](https://developer.mozilla.org/en-US/docs/Learn/Performance/Perceived_performance). This tends to improves user experience by reducing load time frustration and making the page feel more responsive.

Reviewer 4: ```suggestion
The Skeleton Loader displays a placeholder representing the page's content before the page is ready to display. This component is used to improve [perceived performance](https://developer.mozilla.org/en-US/docs/Learn/Performance/Perceived_performance). This tends to improve user experience by reducing load time frustration and making the page feel more responsive.
```

The component is designed to be used in place of the content being loaded. For example:

<LivePreview scope={{SkeletonLoader, Text}} noInline>
  {CompositionExample}
</LivePreview>

### Accessibility

> - The animation can be disabled using the `disabled` prop, or by enabled `prefers-reduced-motion` at the OS level.

Reviewer 3: ```suggestion
- The animation can be disabled using the `disabled` prop, or by enabling the `prefers-reduced-motion` setting at the OS level.
```

> - The animation should be disabled if content takes longer than 5 seconds to load.

Reviewer: I think we should bake this into the animation loop. People won't do it, and potentially won't know.

We can solve it by just making sure the animation loop is 5 seconds on, and 5 seconds off.

Author: Fought with this a bit today, and then realized that the animation was taking almost 5 seconds to fully finish, so I added a small delay before the loop starts.

Reviewer: Looks pretty neat actually, I dig it

> - Skeleton Loader has the `aria-busy` attribute set to `true` by default. This is to indicate that the content is loading. This can be set to `false` using the `disabled` prop.

Reviewer 3: ```suggestion
- Skeleton Loader has the `aria-busy` attribute set to `true` by default. This is to indicate that the content is loading. This is reset to `false` when `disabled` is `true`.
```

- The animation can be disabled by enabling the `prefers-reduced-motion` setting at the OS level.
- Skeleton Loader has the `aria-busy` attribute set to `true` by default. This is to indicate that the content is loading.

## Examples

### Default Skeleton Loader

The default Skeleton Loader can be used to represent a line of text. It uses the `sizeIcon20` token for the `height`, and `borderRadius20` for the `border-radius`. The `width` will be 100% unless a different `width` is specified.

<LivePreview scope={{SkeletonLoader}}>{`<SkeletonLoader />`}</LivePreview>

### Changing the Skeleton Loader height

> Use the `height` prop to change the height of the Skeleton Loader.

Reviewer 4: Would it be useful to say here that the height could be in px _or_ an icon size? It wasn't clear to me on first glance. This could be solved by adding another example that uses icon size for the height. Is there any other type of value one could put into that prop or just those two?

Author: Added some clarifying text.

<LivePreview scope={{SkeletonLoader}}>{`<SkeletonLoader height="150px" />`}</LivePreview>

### Loading text

<LivePreview scope={{Box, Button, SkeletonLoader, Stack, Text}} noInline>

>   {TextExample}

Reviewer 4: I really like how these examples are interactive!!

</LivePreview>

### Loading a paragraph

> 

Reviewer 4: Would it be helpful to mention that using Stack with SkeletonLoader is a common combination for representing lines of text before showing the example?

Author: Added some clarifying text.
